### PR TITLE
[builtins/dict] add -> erase() method

### PR DESCRIPTION
**Pull request by ellen364**
An oddity is that `erase` works with both `->` and `=>` despite only mutating, not transforming. But this behaviour matches other Dict and List methods, so the implementations are at least consistent with each other.

### CODE REVIEW

File: spec/ysh-methods.test.sh [hunk 1]
```diff
@@ -393,6 +393,15 @@ pp line (en2fr => values())
 (List)   ["bonjour","ami","chat"]
 ## END
 
+#### Dict -> erase()
+var book = {title: "The Histories", author: "Herodotus"}
+call book->erase("author")
```
Comment (andychu) [inline spec/ysh-methods.test.sh:398]: I would add another call to test that we do nothing when the key doesn't exist

**Review by andychu — COMMENTED** [review]
Looks great!  Thanks for doing this

Glad the mycpp error message existed!   (Sometimes we don't check for the subset of Python we use)

File: spec/ysh-methods.test.sh (hunk 1 above)
Comment (ellen364) [reply spec/ysh-methods.test.sh:398]: New commit has the extra call.

Doing nothing when the key doesn't exist has really grown on me. Wasn't sure about it at first (felt like a silent failure) but realised that a lot of time I want a dictionary in a particular way and don't care whether the key was there or not, as long as it isn't now.

File: doc/ref/chap-type-method.md [hunk 2]
```diff
@@ -343,6 +343,19 @@ Similar to `keys()`, but returns the values of the dictionary.
 
 ### erase()
 
+Accepts a key and deletes the relevant key-value pair. The method is idempotent.
```
Comment (andychu) [inline doc/ref/chap-type-method.md:346]: How about saying: "erase() ensures that the given key doesn't not exist in the dictionary"

This means either deleting it, or doing nothing if it doesn't exist (which implies that erase() is idempotent)

---

FWIW I believe this same  issue is discussed in Ousterhout's book.  He regretted that for the Tcl language, they made "delete key that does not exist" an error, like it is in Python!

He thinks they should have redefined it to "ensure key does not exist", which is what we're doing here!

https://blog.pragmaticengineer.com/a-philosophy-of-software-design-review/

**Review by andychu — COMMENTED** [review]
small doc suggestion, sorry I didn't see this the first time

I also hint at one reason we differ from Python, which seems to be nice, based on my experience in Oils

**Comment by andychu** [comment]
I'm probably making a release tomorrow ... I may merge this with some tweaks to the docs, to get it in

Thanks for doing this!

File: doc/ref/chap-type-method.md (hunk 2 above)
Comment (ellen364) [reply doc/ref/chap-type-method.md]: The new commit has your suggested phrasing + an additional example. The docs basically repeat the tests now, but I think the extra example helps reinforce what behaviour to expect.

And A Philosophy of Software Design has just gone onto the reading list :slightly_smiling_face:

**Comment by andychu** [comment]
Oh also please keep us updated on Zulip about any code you write with YSH!

Or if you are looking for other things to do, it would be cool to implement `Dict->get() inc() accum()`, which are marked with a red X

I think those are mostly similar to `erase()`, and I'm sure many people will use them!

https://www.oilshell.org/release/0.22.0/doc/ref/toc-ysh.html

**Pull request merged**